### PR TITLE
Updates for latest MOI

### PR DESCRIPTION
**Pull request by odow**
## What it does

In particular:

 - `MOFFile` -> `MOFInstance`
 - Removes `MOFWriter`

Implements:

 - `MOI.write`
 - `MOI.read!`
 - `MOI.copy!`

## `copy!` problems
 - [x] How to match up variables from `src` to variables in `dest`:
    Adopt the convention that:
    ```julia
        src_variables  = MOI.get(src, MOI.ListOfVariableReferences())
        dest_variables = MOI.addvariables!(dest, MOI.get(src, MOI.NumberOfVariables())
        # src_variables[i] <=> dest_variables[i]
    ```
 - [ ] A better way of handling constraints (ref: https://github.com/JuliaOpt/MathOptInterface.jl/issues/163)
It would be nice to go:
```julia
for (F,S) in MOI.get(src, MOI.ListOfConstraints())
    for reference in MOI.get(src, MOI.ListOfConstraintReferences{F,S}())
        func = MOI.get(src, MOI.ConstraintFunction(), src_reference)
        func = rereference!(func, variablemap)
        set  = MOI.get(src, MOI.ConstraintSet(), src_reference)
        if !MOI.canaddconstraint(dest, func, set)
            error("Unable to add the constraint of type ($(F), $(S))")
        end
        MOI.addconstraint!(dest, func, set)
    end
end
```
but this doesn't preserve row-ordering from the `src` model to the `dest` model. The current approach is to define
```julia
MOI.get(m, ::Type{ListOfConstraintReferences})
```
that returns an ordered list of constraint references. This isn't type stable, but preserves ordering.

Once this is done it can just be a generic copy. 

## Example 
The following <s>should work (but is untested)</s> works!!!: cc @mlubin @joaquimg

```julia
using MathOptFormat, MathOptInterfaceGurobi, MathOptInterface
const MOI = MathOptInterface
using Base.Test

# overload as not implemented yet
MOI.canaddconstraint(grb::MathOptInterfaceGurobi.GurobiSolverInstance, f, s) = true

# min 2x
# s.t. x >= 3
mof = MathOptFormat.MOFInstance()
x = MOI.addvariable!(mof)
MOI.addconstraint!(mof, MOI.SingleVariable(x), MOI.GreaterThan(3.0))
MOI.set!(mof, MOI.ObjectiveSense(), MOI.MinSense)
MOI.set!(mof, MOI.ObjectiveFunction(), MOI.ScalarAffineFunction([x], [2.0], 0.0))
MOI.write(mof, "model.mof.json")

mof2 = MathOptFormat.MOFInstance()
MOI.read!(mof2, "model.mof.json")
grb = MOI.SolverInstance(MOIGurobiSolver())
MOI.copy!(grb, mof2)


MOI.optimize!(grb)
v = MOI.get(grb, MOI.ListOfVariableReferences())
@test MOI.get(grb, MOI.VariablePrimal(), v[1]) == 3.0
@test MOI.get(grb, MOI.ObjectiveValue()) == 6.0
```

### CODE REVIEW

File: src/reader.jl [hunk 1]
```diff
@@ -1,48 +1,73 @@
...
+    if haskey(dict, str) && MOI.canset(dest, attr, ref)
+        MOI.set!(dest, attr, ref, dict[str])
+    end
+end
 
-Create a new MathOptInterface solver instance using `solver` from the MOFFile
-located at the path `file`.
-"""
-MOI.SolverInstance(file::String, solver) = MOI.SolverInstance(MOFFile(file), solver)
+function MOI.copy!(dest::MOI.AbstractInstance, src::MOFInstance)
```
Comment (mlubin) [inline src/reader.jl]: I can see some ambiguity issues if an instance defines ``MOI.copy!(dest::MyInstance, src::MOI.AbstractInstance)``. Can we just have one generic version of `MOI.copy` defined in MOI or MOIU?

Comment (odow) [reply src/reader.jl]: I would prefer a convention that only `src` instances defined `copy!` and the 2nd argument was concretely typed.

But I suppose you could make one generic version that checked `canget` and `canset` for all attributes and constraints.

**Comment by mlubin** [comment]
What's the purpose of `MOFWriter`?

**Comment by odow** [comment]
> What's the purpose of `MOFWriter`?

Legacy of wanting to go `m = MOI.SolverInstance(MOFWriter())` instead of `m = MOFInstance()`

**Comment by mlubin** [comment]
Drop the legacy and drop the `m` :)

**Comment by odow** [comment]
> What's the purpose of `MOFWriter`?

Actually it was `supportsproblem`. But that should probably just be changed to `supportsproblem(::AbstractInstance, ...)` rather than `AbstractSolver`.

**Comment by mlubin** [comment]
That actually opens up an interesting can of worms. It's a bit inconsistent that we don't have something analogous to `AbstractSolver` for standalone instances.

**Comment by odow** [comment]
Maybe `solver` is a misnomer and they should be called
 - InstanceCreator
 - Instantiator
 - InstanceGenerator
 - InstanceController

`MOI.Instance(::InstanceCreator)` builds either a `StandaloneInstance` or a `SolverInstance`

**Pull request merged**